### ISSUE TITLE
Confustion about predictive structure-wise variance

### ISSUE THREAD

**Issue by ThisGame42** [yuta-hi/bayesian_unet#3]
Dear @yuta-hi ,

I'm currently reading your paper and I must say it is indeed amazing work. However, I personally find the part in your paper where you describe predictive structure-wise variance (PSR) a little difficult to understand. That is, this bit:

![image](https://user-images.githubusercontent.com/5229138/83936354-4b9be700-a806-11ea-908c-453f4fde3daf.png)

My understanding is that for each pixel in the output, we want to sum over the variances of it being each class, and then sum all the variances. For example, if we have two classes (0,1), and our output size is 1x2 (so that we have output pixels: y_00, y_01), what we want to do is to sum var(y_00 = class_0) + var(y_00 = class 1) + var(y_01 = class_0) + var(y_01 = class_1) and then normalise the sum by the size of the number of pixels in the output.

Can you please kindly let me know if my understanding is correct? I apologise if this question seems a bit silly to you as I've just begun to march into the world of deep learning. 

Thank you very much for your time reading this through and I look forward to seeing your reply.

Kind regards,
JZ

**Comment by yuta-hi** [comment]
The point of this equation is to calculate the uncertainty per unit volume.
Here's a simple code for it.

```python
n_class = 10

predictor = BayesianUNet(ndim=2,
                         out_channels=n_class)

model = MCSampler(predictor,
                  mc_iteration=args.mc_iteration,
                  activation=partial(F.softmax, axis=1),
                  reduce_mean=partial(F.argmax, axis=1),
                  reduce_var=partial(F.mean, axis=1)) # NOTE: Average the predicted variance of each class

infer = Inferencer(test_iter, model, device=args.gpu[0])

label, variance = infer.run()

# NOTE: 
# label.shape: (number of slices, 512, 512)
# variance.shape: (number of slices, 512, 512)

# convert the pixel-wise variance to the structure-wise variance
for i in range(1, n_class): # Zero is assumed to be the background.
    mask = label == i
    psv = np.mean(variance[mask]) # NOTE: Normalize to the variance per unit volume
    print('PSV (#%02d) = %f' % (i, psv))
```

**Comment by ThisGame42** [comment]
Hi @yuta-hi thanks very much for this detailed explanation. Now I understand how PSV is derived. Much appreciated!

Closing this issue as well. Have a good day!
